File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        
        dummy = ListNode(0, head)
        left = dummy
        right = dummy
        
        while n > 0 and right:
            right = right.next
            n -= 1
        
        for i in range(n):
            right = right.next
        
        
        while (right.next != None):
            left = left.next
            right = right.next
        
        left.next = left.next.next
        
        return dummy.next
            
            

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        int i = 0;
        
        ListNode* current = head;
        
        while (current != NULL) {
            i++;
            current = current->next;
        }
        
        if (i == n) {
            return head->next;
        }
        
        current = head;
        
        while (i != n+1) {
            current = current->next;
            i--;
        }
        
        current->next = current->next->next;
        
        return head;
        
        
        
        
    }
};